Marlin/src/core/serial_hook.h

```diff
@@ -24,8 +24,32 @@
 #include "macros.h"
 #include "serial_base.h"
 
-// Used in multiple places
-typedef int8_t serial_index_t;
+
+// A mask containing a bitmap of the serial port to act upon
+// This is written to ensure a serial index is never used as a serial mask
+class SerialMask
+{
+  uint8_t mask;
+
+  // This constructor is private to ensure you can't convert an index to a mask
+  // The compiler will stop here if you are mixing index and mask in your code.
+  // If you need to, you'll have to use the explicit static "from" method here
+  SerialMask(const serial_index_t);
+
+public:
+  inline constexpr bool enabled(const SerialMask PortMask) const    { return mask & PortMask.mask; }
+  inline constexpr SerialMask combine(const SerialMask other) const { return SerialMask(mask | other.mask); }
+  inline constexpr SerialMask operator<< (const int offset) const   { return SerialMask(mask << offset); }
+  static inline SerialMask from(const serial_index_t index) {
+    if (index.valid()) return SerialMask(_BV(index.index));
+    return SerialMask(0); // A invalid index mean no output
+  }
+
+  constexpr SerialMask(const uint8_t mask) : mask(mask) {}
+  constexpr SerialMask(const SerialMask & other) : mask(other.mask) {} // Can't use = default here since not all framework support this
+
+  static constexpr uint8_t All = 0xFF;
+};
```

Comment:
I'm really fan of everything as type... I mean, I love Ruby, that everything is an objet... But isn't a bit of overkill treat a _mask_ as a full class (even if it's `const expr`)?

Reply:
Of course it is. :wink: I wanted to check if there was place we'd missed by mixing serial_index and mask. I didn't want to do it by hand, and the compiler is the most dumbest tool to do that for me. So, yes it's overkill but it's useful. The compiler directly screamed on the autoreport class that was mixing stuff up (storing a mask in a serial_index_t member). With the SERIAL_ALL set to -1, this would have silented the autoreport since serial_index_t == -1 is the value to silent the output

Reply:
BTW, I can remove that now that I've build all the code with it and found the errors. I can define it back to uint8 instead, but that would prevent detecting future "mix" bug. Let me know what you prefer.

Reply:
No no, keep. It was just a comment, not a critic :-)

Reply:
This is a fine approach and makes the system more robust. I would like to see more fool-proofing applied in various systems, but as they say… it's hard to find good help.

I'm working on applying a simple `celsius_t` type throughout Marlin to make it clear where celsius temperatures are being referenced, and then I plan to refactor the heater PID loop into a template class because it is very repetitive. I also have a project posted to create a Feature class and make Features take some responsibility for their own EEPROM-stored settings using something like a "delegate" pattern. All these changes will benefit from a little extra bullet-proofing.

We've been discussing "asserts" to catch bad function parameters and using other conditional guards to catch bugs in a test / debug build. And we have a simulation that can attach directly to the debugger and be used to catch bugs more quickly by running G-code at thousand-times speed. These tools together with "typo prevention" will go a long way to making a more stable Marlin overall.

Reply:
Sure, but for simple things I would use just `using` to create a type alias.
And the methods would be just const expr functions, not needing a class at all, as there’s no state to hold.
But the code is great. The mask and index logic are very well encapsulated now. And I suppose the the compiler will be clever to generate a optimized code.